### ISSUE TITLE
UnderC does not build

### ISSUE THREAD

**Issue by belliash** [stevedonovan/UnderC#1]
Hey, i got many errors while trying to build UnderC:

g++ -c -O1 -Wno-deprecated  -Wno-write-strings  -D_CONSOLE -D_USE_READLINE tparser.cpp
In file included from class.h:6:0,
                 from types.h:155,
                 from common.h:6,
                 from parser.y:18:
table.h: In member function ‘int Allocator::size()’:
table.h:42:18: error: cast from ‘char_’ to ‘int’ loses precision [-fpermissive]
table.h:42:32: error: cast from ‘char_’ to ‘int’ loses precision [-fpermissive]
table.h: In member function ‘int Allocator::offset(void_)’:
table.h:63:17: error: cast from ‘void_’ to ‘int’ loses precision [-fpermissive]
table.h:63:28: error: cast from ‘char_’ to ‘int’ loses precision [-fpermissive]
parser.y: In function ‘int yyparse()’:
parser.y:675:48: warning: cast to pointer from integer of different size [-Wint-to-pointer-cast]
parser.y:676:79: warning: cast to pointer from integer of different size [-Wint-to-pointer-cast]
make: *_\* [tparser.o] Error 1

I tried to add -fpermissive to CFLAGS, but then i got this:

g++ -c -O1 -Wno-deprecated  -Wno-write-strings -fpermissive  -D_CONSOLE -D_USE_READLINE tparser.cpp
In file included from class.h:6:0,
                 from types.h:155,
                 from common.h:6,
                 from parser.y:18:
table.h: In member function ‘int Allocator::size()’:
table.h:42:18: warning: cast from ‘char_’ to ‘int’ loses precision [-fpermissive]
table.h:42:32: warning: cast from ‘char_’ to ‘int’ loses precision [-fpermissive]
table.h: In member function ‘int Allocator::offset(void_)’:
table.h:63:17: warning: cast from ‘void_’ to ‘int’ loses precision [-fpermissive]
table.h:63:28: warning: cast from ‘char_’ to ‘int’ loses precision [-fpermissive]
parser.y: In function ‘int yyparse()’:
parser.y:675:48: warning: cast to pointer from integer of different size [-Wint-to-pointer-cast]
parser.y:676:79: warning: cast to pointer from integer of different size [-Wint-to-pointer-cast]
g++ -c -O1 -Wno-deprecated  -Wno-write-strings -fpermissive  -D_CONSOLE -D_USE_READLINE function_match.cpp
In file included from class.h:6:0,
                 from types.h:155,
                 from function_match.h:6,
                 from function_match.cpp:8:
table.h: In member function ‘int Allocator::size()’:
table.h:42:18: warning: cast from ‘char_’ to ‘int’ loses precision [-fpermissive]
table.h:42:32: warning: cast from ‘char_’ to ‘int’ loses precision [-fpermissive]
table.h: In member function ‘int Allocator::offset(void_)’:
table.h:63:17: warning: cast from ‘void_’ to ‘int’ loses precision [-fpermissive]
table.h:63:28: warning: cast from ‘char_’ to ‘int’ loses precision [-fpermissive]
g++ -c -O1 -Wno-deprecated  -Wno-write-strings -fpermissive  -D_CONSOLE -D_USE_READLINE class.cpp
In file included from class.h:6:0,
                 from types.h:155,
                 from common.h:6,
                 from class.cpp:13:
table.h: In member function ‘int Allocator::size()’:
table.h:42:18: warning: cast from ‘char_’ to ‘int’ loses precision [-fpermissive]
table.h:42:32: warning: cast from ‘char_’ to ‘int’ loses precision [-fpermissive]
table.h: In member function ‘int Allocator::offset(void_)’:
table.h:63:17: warning: cast from ‘void_’ to ‘int’ loses precision [-fpermissive]
table.h:63:28: warning: cast from ‘char_’ to ‘int’ loses precision [-fpermissive]
class.cpp: In member function ‘void Class::set_base_class(Class_, int)’:
class.cpp:77:41: warning: cast to pointer from integer of different size [-Wint-to-pointer-cast]
g++ -c -O1 -Wno-deprecated  -Wno-write-strings -fpermissive  -D_CONSOLE -D_USE_READLINE code.cpp
In file included from class.h:6:0,
                 from types.h:155,
                 from common.h:6,
                 from code.cpp:8:
table.h: In member function ‘int Allocator::size()’:
table.h:42:18: warning: cast from ‘char_’ to ‘int’ loses precision [-fpermissive]
table.h:42:32: warning: cast from ‘char_’ to ‘int’ loses precision [-fpermissive]
table.h: In member function ‘int Allocator::offset(void_)’:
table.h:63:17: warning: cast from ‘void_’ to ‘int’ loses precision [-fpermissive]
table.h:63:28: warning: cast from ‘char_’ to ‘int’ loses precision [-fpermissive]
code.cpp: In member function ‘void UCContext::emit_reference(PExpr, int, int)’:
code.cpp:211:51: warning: cast to pointer from integer of different size [-Wint-to-pointer-cast]
code.cpp:212:52: warning: cast from ‘FBlock_’ to ‘int’ loses precision [-fpermissive]
g++ -c -O1 -Wno-deprecated  -Wno-write-strings -fpermissive  -D_CONSOLE -D_USE_READLINE common.cpp
In file included from class.h:6:0,
                 from types.h:155,
                 from common.h:6,
                 from common.cpp:10:
table.h: In member function ‘int Allocator::size()’:
table.h:42:18: warning: cast from ‘char_’ to ‘int’ loses precision [-fpermissive]
table.h:42:32: warning: cast from ‘char_’ to ‘int’ loses precision [-fpermissive]
table.h: In member function ‘int Allocator::offset(void_)’:
table.h:63:17: warning: cast from ‘void_’ to ‘int’ loses precision [-fpermissive]
table.h:63:28: warning: cast from ‘char_’ to ‘int’ loses precision [-fpermissive]
common.cpp: In function ‘Label_ Parser::create_goto_label(char_)’:
common.cpp:862:27: warning: cast from ‘Label_’ to ‘int’ loses precision [-fpermissive]
common.cpp: In function ‘Label\* Parser::check_goto_label(PEntry)’:
common.cpp:872:31: warning: cast to pointer from integer of different size [-Wint-to-pointer-cast]
common.cpp: In member function ‘void Parser::ParserState::add_namespace(string)’:
common.cpp:1073:22: warning: cast from ‘Namespace_’ to ‘int’ loses precision [-fpermissive]
common.cpp:1078:31: warning: cast to pointer from integer of different size [-Wint-to-pointer-cast]
common.cpp: In function ‘FunctionEntry_ Parser::create_function_entry(Signature_, const string&, PEntry)’:
common.cpp:1195:20: warning: cast from ‘FunctionEntry_’ to ‘int’ loses precision [-fpermissive]
common.cpp: In function ‘FunctionEntry\* Parser::lookup_function_entry(Signature_, const string&, bool, Function__, int&)’:
common.cpp:1237:32: warning: cast to pointer from integer of different size [-Wint-to-pointer-cast]
common.cpp:1262:43: warning: cast to pointer from integer of different size [-Wint-to-pointer-cast]
common.cpp: In member function ‘Function_ Parser::ParserState::declare_function(Type, const string&, PExpr)’:
common.cpp:1427:48: warning: cast to pointer from integer of different size [-Wint-to-pointer-cast]
common.cpp: In function ‘void Parser::do_start_try_block()’:
common.cpp:1995:64: warning: cast from ‘CatchHandler_’ to ‘int’ loses precision [-fpermissive]
common.cpp: In function ‘bool Parser::expression_end(PExpr)’:
common.cpp:2115:66: warning: cast from ‘PEntry {aka Entry_}’ to ‘int’ loses precision [-fpermissive]
common.cpp: In function ‘Namespace\* Parser::lookup_namespace(char_)’:
common.cpp:2210:56: warning: cast to pointer from integer of different size [-Wint-to-pointer-cast]
common.cpp: In function ‘void Parser::dump_function(PEntry, bool, int, std::ostream&)’:
common.cpp:2358:46: warning: cast to pointer from integer of different size [-Wint-to-pointer-cast]
g++ -c -O1 -Wno-deprecated  -Wno-write-strings -fpermissive  -D_CONSOLE -D_USE_READLINE directcall.cpp
In file included from class.h:6:0,
                 from types.h:155,
                 from common.h:6,
                 from directcall.cpp:9:
table.h: In member function ‘int Allocator::size()’:
table.h:42:18: warning: cast from ‘char_’ to ‘int’ loses precision [-fpermissive]
table.h:42:32: warning: cast from ‘char_’ to ‘int’ loses precision [-fpermissive]
table.h: In member function ‘int Allocator::offset(void_)’:
table.h:63:17: warning: cast from ‘void_’ to ‘int’ loses precision [-fpermissive]
table.h:63:28: warning: cast from ‘char_’ to ‘int’ loses precision [-fpermissive]
directcall.cpp: In function ‘int Builtin::convert_ordinal(char_)’:
directcall.cpp:1026:23: warning: format ‘%x’ expects argument of type ‘unsigned int_’, but argument 3 has type ‘long unsigned int_’ [-Wformat]
directcall.cpp: Assembler messages:
directcall.cpp:224: Error: invalid instruction suffix for `push'
directcall.cpp:229: Error: operand type mismatch for`call'
make: *_\* [directcall.o] Error 1

My gcc -v:
COLLECT_GCC=/usr/x86_64-pc-linux-gnu/gcc-bin/4.6.2/gcc
COLLECT_LTO_WRAPPER=/usr/libexec/gcc/x86_64-pc-linux-gnu/4.6.2/lto-wrapper
Target: x86_64-pc-linux-gnu
Configured with: /var/tmp/portage/sys-devel/gcc-4.6.2/work/gcc-4.6.2/configure --prefix=/usr --bindir=/usr/x86_64-pc-linux-gnu/gcc-bin/4.6.2 --includedir=/usr/lib/gcc/x86_64-pc-linux-gnu/4.6.2/include --datadir=/usr/share/gcc-data/x86_64-pc-linux-gnu/4.6.2 --mandir=/usr/share/gcc-data/x86_64-pc-linux-gnu/4.6.2/man --infodir=/usr/share/gcc-data/x86_64-pc-linux-gnu/4.6.2/info --with-gxx-include-dir=/usr/lib/gcc/x86_64-pc-linux-gnu/4.6.2/include/g++-v4 --host=x86_64-pc-linux-gnu --build=x86_64-pc-linux-gnu --disable-altivec --disable-fixed-point --with-ppl --with-cloog --disable-ppl-version-check --with-cloog-include=/usr/include/cloog-ppl --enable-lto --enable-nls --without-included-gettext --with-system-zlib --disable-werror --enable-secureplt --enable-multilib --enable-libmudflap --disable-libssp --enable-libgomp --with-python-dir=/share/gcc-data/x86_64-pc-linux-gnu/4.6.2/python --enable-checking=release --disable-libgcj --disable-libquadmath --enable-languages=c,c++ --enable-shared --enable-threads=posix --enable-__cxa_atexit --enable-clocale=gnu --enable-targets=all --with-bugurl=http://bugs.gentoo.org/ --with-pkgversion='Gentoo 4.6.2 p1.4, pie-0.5.0'
Thread model: posix
gcc version 4.6.2 (Gentoo 4.6.2 p1.4, pie-0.5.0)

**Comment by stevedonovan** [comment]
Yeah, sorry about the number of warnings that still need to be cleared up. But your basic problem is that UC will currently only build on x86 - there is some inline assembler in directcall.cpp.  I intend to replace this custom code with libffi and then we can get portable.

**Comment by belliash** [comment]
I have added -m32 to CFLAGS which forces compiler to produce 32bit code but i came across another problem, this time with readline:

g++ -c -O1 -m32 -Wno-deprecated  -Wno-write-strings -fpermissive  -D_CONSOLE -D_USE_READLINE tparser.cpp
g++ -c -O1 -m32 -Wno-deprecated  -Wno-write-strings -fpermissive  -D_CONSOLE -D_USE_READLINE function_match.cpp
g++ -c -O1 -m32 -Wno-deprecated  -Wno-write-strings -fpermissive  -D_CONSOLE -D_USE_READLINE class.cpp
g++ -c -O1 -m32 -Wno-deprecated  -Wno-write-strings -fpermissive  -D_CONSOLE -D_USE_READLINE code.cpp
g++ -c -O1 -m32 -Wno-deprecated  -Wno-write-strings -fpermissive  -D_CONSOLE -D_USE_READLINE common.cpp
g++ -c -O1 -m32 -Wno-deprecated  -Wno-write-strings -fpermissive  -D_CONSOLE -D_USE_READLINE directcall.cpp
directcall.cpp: In function ‘int Builtin::convert_ordinal(char_)’:
directcall.cpp:1026:23: warning: format ‘%x’ expects argument of type ‘unsigned int_’, but argument 3 has type ‘long unsigned int_’ [-Wformat]
g++ -c -O1 -m32 -Wno-deprecated  -Wno-write-strings -fpermissive  -D_CONSOLE -D_USE_READLINE dissem.cpp
g++ -c -O1 -m32 -Wno-deprecated  -Wno-write-strings -fpermissive  -D_CONSOLE -D_USE_READLINE engine.cpp
g++ -c -O1 -m32 -Wno-deprecated  -Wno-write-strings -fpermissive  -D_CONSOLE -D_USE_READLINE expressions.cpp
g++ -c -O1 -m32 -Wno-deprecated  -Wno-write-strings -fpermissive  -D_CONSOLE -D_USE_READLINE function.cpp
g++ -c -O1 -m32 -Wno-deprecated  -Wno-write-strings -fpermissive  -D_CONSOLE -D_USE_READLINE breakpoints.cpp
g++ -c -O1 -m32 -Wno-deprecated  -Wno-write-strings -fpermissive  -D_CONSOLE -D_USE_READLINE tokens.cpp
In file included from /usr/include/readline/readline.h:36:0,
                 from tokens.cpp:10:
/usr/include/readline/rltypedefs.h:64:28: error: typedef ‘rl_getc_func_t’ is initialized (use decltype instead)
/usr/include/readline/rltypedefs.h:64:28: error: ‘FILE’ was not declared in this scope
/usr/include/readline/rltypedefs.h:64:28: error: expected primary-expression before ‘)’ token
In file included from tokens.cpp:10:0:
/usr/include/readline/readline.h:419:20: error: ‘FILE’ was not declared in this scope
/usr/include/readline/readline.h:419:20: error: expected primary-expression before ‘)’ token
/usr/include/readline/readline.h:542:8: error: ‘FILE’ does not name a type
/usr/include/readline/readline.h:543:8: error: ‘FILE’ does not name a type
/usr/include/readline/readline.h:565:8: error: ‘rl_getc_func_t’ does not name a type
/usr/include/readline/readline.h:867:3: error: ‘FILE’ does not name a type
/usr/include/readline/readline.h:868:3: error: ‘FILE’ does not name a type
make: *_\* [tokens.o] Error 1

**Comment by stevedonovan** [comment]
Can you try this?

 make NO_READLINE=1

**Comment by belliash** [comment]
Yes, it built fine then.
Looks like you should disable readline and force -m32 flag to let it build on 64bit host linux.

But now i got segmentation fault when i try to output anything, using both printf and cout. I know it contains a lot of trash, but shows the problem. I will try to rebuild it with debug enabled at a later point.

$ ./ucc
UnderC C++ Interpreter vs 1.3.1L
Steve Donovan, 2001-2003
This program is GPL'd; see LICENCE for details
CON 0:Cannot open defs.h
Now opening defs.h in UC home directory
(1) access violation
;> printf("gfgdfgd");
Segmentation fault

$ linux32 ./ucc
UnderC C++ Interpreter vs 1.3.1L
Steve Donovan, 2001-2003
This program is GPL'd; see LICENCE for details
CON 0:Cannot open defs.h
Now opening defs.h in UC home directory
(1) access violation
;> u
;; 
;; ;
CON 3:syntax error, unexpected ';', expecting ':'
CON 3:Cannot find 'u'
;> printf("gfgf")
;; ;
Segmentation fault

$ ./ucc
UnderC C++ Interpreter vs 1.3.1L
Steve Donovan, 2001-2003
This program is GPL'd; see LICENCE for details
CON 0:Cannot open defs.h
Now opening defs.h in UC home directory
(1) access violation
;> x=5;
CON 1:syntax error, unexpected ASSIGN, expecting ':'
CON 1:Cannot find 'x'
;> int x=5;
;> cout << x;
Segmentation fault

$ ./ucc
UnderC C++ Interpreter vs 1.3.1L
Steve Donovan, 2001-2003
This program is GPL'd; see LICENCE for details
CON 0:Cannot open defs.h
Now opening defs.h in UC home directory
(1) access violation
;> int x = 5;
;> x++;
(int) 5
;> ^[[A
CON 3:syntax error, unexpected $undefined
CON 3:error in expression
;> x++;
(int) 6
;> ++x;
(int) 8
;> printf
;; ;  
1 int printf(const char\* )
;> printf();
Candidates were
1 int printf(const char\* )
CON 8:Function cannot match 0 parameters
;> printf("");
Segmentation fault

**Comment by belliash** [comment]
hmmm,
I have just rebuilt it with debug enabled (make DEBUG=1) and the problem is gone. Maybe some trash was added by compiler when i was playing with flags.

**Comment by belliash** [comment]
No, sorry for multiple posts, but i have done make clean && make and the problem appeared again:

$ ./ucc
UnderC C++ Interpreter vs 1.3.1L
Steve Donovan, 2001-2003
This program is GPL'd; see LICENCE for details
CON 0:Cannot open defs.h
Now opening defs.h in UC home directory
(1) access violation
;> cout << "gfdgfddgf";
Segmentation fault

**Comment by stevedonovan** [comment]
The DEBUG=1 success reminds me of a problem we had recently getting the old codebase to work on x86.  The inline assembler was overwriting registers that the optimized code needed.  I saved the values of ECX and EBX and the problem went away.  Looks like I'm going to have to look at the x64 host problem!

The real solution is to find a more portable solution to the FFI problem, which is what libffi will bring to the party. That assembler stuff is very fragile!

**Comment by belliash** [comment]
The -O1 cflag couses descrbed above problem. It is enabled only when there is no debug enabled what clarifies why it does not work only with debug disabled. Removing it from cflags solves the issue.

$ ./ucc
UnderC C++ Interpreter vs 1.3.1L
Steve Donovan, 2001-2003
This program is GPL'd; see LICENCE for details
CON 0:Cannot open defs.h
Now opening defs.h in UC home directory
;> COUT << 5;
CON 1:syntax error, unexpected LSHIFT, expecting ':'
CON 1:Cannot find 'COUT'
;> cout << 5;
5(std::ostream&) ostream {}
;> printf("gfgdfgd"); 
gfgdfgd(int) 7

To sum up it was necessary to add -m32 and remove -O1 & readline to make it working correctly.

**Comment by stevedonovan** [comment]
Well, I'm glad you succeeded!  It will do for now, but I think the priority now is a portable build with no fancy tricks.  I've recently come back to the codebase after a break of nearly 10 years, and there's lots of little things about the code that irritates me. I keep thinking of exciting things like implementing C++11 features, but what the beast needs is a good cross-platform, clean build.  In the process I'll clean up this nonsense of using 'char *' when I mean 'const char *'.  Sometimes we are embarrassed by our past ;)

**Comment by belliash** [comment]
I totally agree, just let You know about my observation that might be used as a temporary hack allowing more people to try it without complaining ;)

BTW: Have you also thought about 2nd binary called eg ucc-cgi that would execute code between <% %> and output everything else to stdout? This would allow to easily create web pages using UnderC as it does not really need any special tricks to work. Actually putting HTML into printf or cout is a painful solution ;) I know it is not intended to interpret web pages, but this would be a great enhancement and a feature anyway.

**Comment by stevedonovan** [comment]
Good point!  It will be easy enough to add your changes for x64, perhaps with a little warning at the end. Thank you for your patience.

You're right, ucc-cgi is perfectly possible, although its major use would be quick prototyping. Once things work, we have compilers ;)   Although thinking about <%, the major win is that much of the page is HTML, with inline C++ for dynamic template expansion.

**Comment by belliash** [comment]
Thats right, this would allow both writing web pages and messing HTML with C++ like its done in PHP or ASP.NET. That would be much easier to work with than using specialized framework and i think shouldn't require a lot of work. Plz correct me if i am wrong.

**Comment by stevedonovan** [comment]
I must think about how to do it. As always, my problem is priorities - what happens next? And my feeling is that 'portability' is going to be the next big step.

**Comment by stevedonovan** [comment]
Interesting; exactly the same problem here on Windows with gcc 4.6.1 - crash on -O1.

Works with -O1 when -fno-omit-frame-pointer is specified (since this default changed with 4.6).  I'm going to try to just apply this option to the 'delicate' code in directcall.cpp.

As for the readline issue, I am thinking that maybe the linenoise project would be a better fit ?

**Comment by belliash** [comment]
I have never heard about linenoise and personally would stick with readline. It is one of most popular, just wonder if it is really necessary, im not sure what kind of code depends on it, however looks like everything works fine w/o it. I have not looked thorough at it but at first impression looks like newer version of readline introduces some changes that are incompatible with older version, afaik it complained about FILE typedef, maybe even just missing header? Im actually a little bit busy, but can do some test in free time ;)

BTW I found another problem:

UnderC C++ Interpreter vs 1.3.1L
Steve Donovan, 2001-2003
This program is GPL'd; see LICENCE for details
CON 0:Cannot open defs.h
Now opening defs.h in UC home directory
;> exit
;; exit 1;
CON 2:Cannot find 'exit'
CON 2:syntax error, unexpected TOKEN, expecting ':'
CON 2:Cannot find 'exit'
;> exit(1);
CON 3:syntax error, unexpected '(', expecting ':'
CON 3:Cannot find 'exit'
;> exit:
Segmentation fault

This should rather produce error instead of causing segmentation fault ;) I cannot try to reproduce it on x86, only x86_64.

**Comment by stevedonovan** [comment]
The latest push modifies the flags for directcall.cpp so that it does create a stack frame for these delicate functions; works fine with my mingw gcc 4.6. Also explicitly has -m32, so it should work fine.  In fact, I've shipped linenoise with this one, because it's really very small (about 650 lines) and then the resulting binary now only depends on libstdc++.  It's very irritating having an interactive prompt without history and basic line editing, and I think it is a practical solution to your readline problems.

Yes, sometimes UC can be too delicate with compile errors. I can't quite repeat what you did, it doesn't segfault, but then this is a Windows MSVC build.

**Comment by belliash** [comment]
Any news about this issue?
